### DISCUSSION TITLE
preset-ext and strict mode question

### DISCUSSION THREAD

**Discussion by zslabs** [tw-in-js/twind#302] [Q&A] [answered]
Hey there! As always, really appreciate all the hard-work going into twind v1. I was looking into two aspects of the current beta and was hoping to get a bit of clarification:

1. Is [preset-ext](https://github.com/tw-in-js/twind/tree/next/packages/preset-ext) included by default in [preset-tailwind](https://github.com/tw-in-js/twind/tree/next/packages/preset-tailwind)? My hope was to deter the usage of these magic-values `p-[4.5343px]` and depend more on what is in the theme config, but I was having trouble finding if this in the tailwind preset folder.
2. Is `strict` mode going to be a thing in v1? Along the same idea of removing the ability to create magic-values/selectors, `strict` mode was nice in `0.X.X` to prevent incorrect usage of the classes. An additional feature would be the ability to whitelist some REGEX classes for 3rd party entities.

**Comment by sastan** [comment, marked as answer]
> Hey there! As always, really appreciate all the hard-work going into twind v1. I was looking into two aspects of the current beta and was hoping to get a bit of clarification:
> 
> 1. Is [preset-ext](https://github.com/tw-in-js/twind/tree/next/packages/preset-ext) included by default in [preset-tailwind](https://github.com/tw-in-js/twind/tree/next/packages/preset-tailwind)? My hope was to deter the usage of these magic-values `p-[4.5343px]` and depend more on what is in the theme config, but I was having trouble finding if this in the tailwind preset folder.

This is a builtin feature of twind. If a rule uses the `fromTheme` helper it automatically supports arbitrary values. Here is the example for padding: https://github.com/tw-in-js/twind/blob/next/packages/preset-tailwind/src/rules.ts#L246. The basic idea is if a value is not found in the theme we try to check if it is an arbitrary value (https://github.com/tw-in-js/twind/blob/next/packages/twind/src/rules.ts#L63).

> 2. Is `strict` mode going to be a thing in v1? Along the same idea of removing the ability to create magic-values/selectors, `strict` mode was nice in `0.X.X` to prevent incorrect usage of the classes. An additional feature would be the ability to whitelist some REGEX classes for 3rd party entities.

Something like that will be added. I'm still thinking about the DX for this.

**Reply by zslabs** [reply]
Thanks for the reply and sounds good! When `strict` mode is investigated a bit more, it sounds like that might be a good opportunity for me to look at adding a PR that could adjust those rules. Or -- maybe I could spin up an eslint plugin to forbid brackets 🤔 . Either way, thanks again and have a great day!
